File: bindings/java/src/org/raml/amf/utils/Clojure.java
package org.raml.amf.utils;

import clojure.lang.*;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by antoniogarrote on 04/05/2017.
 */

/**
 * Clojure interop utilties.
 * Many of this utilities and this particular approach to interop is because of a bug related to core.async preventing
 * us to use aot compilation and the class that could have been generated in api-modeling-framework.core
 */
public class Clojure {

    public static final String CLOJURE_CORE = "clojure.core";
    public static final String API_MODELING_FRAMEWORK_CORE = "api-modeling-framework.core";
    public static final String API_MODELING_FRAMEWORK_MODEL_DOCUMENT = "api-modeling-framework.model.document";
    public static final String API_MODELING_FRAMEWORK_MODEL_DOMAIN = "api-modeling-framework.model.domain";
    public static final String CHESHIRE_CORE = "cheshire.core";
    public static final Var REQUIRE= RT.var("clojure.core", "require");


    public static Object require(String nsName) {
        return REQUIRE.invoke(Symbol.intern(nsName));
    }

    /**
     * Looks up a var by name in the given namespace.
     *
     * The var can subsequently be invoked if it is a function.
     * @param nsName
     * @param varName
     * @return
     */
    public static Var var(String nsName, String varName) {
        return RT.var(nsName,varName);
    }

    public static Keyword kw(String name) {
        return Keyword.find(name);
    }

    public static Object getKw(Object target, String name) {
        IFn getFn= var(CLOJURE_CORE, "get");
        return getFn.invoke(target, kw(name));
    }

    public static Object setKw(Object target, String name, Object value) {
        IFn setFn= var(CLOJURE_CORE, "assoc");
        return setFn.invoke(target, kw(name), value);
    }

    public static IPersistentVector list(List list) {
        IPersistentVector tmp = PersistentVector.EMPTY;
        for(Object e : list) {
            tmp = tmp.cons(e);
        }

        return tmp;
    }

    public static <T> List<T> toJavaList(List xs) {
        ArrayList<T> tmp = new ArrayList<>();
        IFn firstFn= var(CLOJURE_CORE, "first");
        IFn restFn= var(CLOJURE_CORE, "rest");
        IFn emptyFn= var(CLOJURE_CORE, "empty?");

        while(! (Boolean) emptyFn.invoke(xs)) {
            tmp.add((T) firstFn.invoke(xs));
            xs = (List) restFn.invoke(xs);
        }

        return tmp;
    }

    public static IPersistentMap map() {
        return PersistentHashMap.EMPTY;
    }
}

File: bindings/js/src/Clojure.ts
/**
 * Created by antoniogarrote on 05/05/2017.
 */

require("api-modeling-framework");
const cljscore = global['cljs'].core;

export class Clojure {
    public static kw(s: string): any {
        return cljscore.keyword(s);
    }

    public static cljsToJs(o: any): any {
        return cljscore.clj__GT_js(o);
    }

    public static jsToCljs(o: any): any {
        return cljscore.js__GT_clj(o);
    }

    public static cljsMap<T>(xs: any, f:(x: any) => T): T[] {
        const acc: T[] = [];
        while(! cljscore.empty_QMARK_(xs)) {
            const first = cljscore.first(xs);
            xs = cljscore.rest(xs);
            acc.push(f(first));
        }

        return acc;
    }

    public static toCljsSeq(xs: any[]): any {
        let acc = cljscore.vec();
        xs.forEach(x => {
            acc = cljscore.cons(x, acc);
        });

        return cljscore.reverse(acc);
    }

    public static core = global['cljs'].core;
    public static amf = global['api_modeling_framework'].core;
    public static amf_document = global['api_modeling_framework'].model.document;
    public static amf_domain = global['api_modeling_framework'].model.domain;
}